overlord/state/warning.go

```diff
@@ -0,0 +1,280 @@
+	message string
+	// the first time one of these messages was created
+	firstAdded time.Time
+	// the last time one of these was created
+	lastAdded time.Time
+	// the last time one of these was shown to the user
+	lastShown time.Time
+	// how much time since one of these was last added should we drop the message
+	expireAfter time.Duration
+	// how much time since one of these was last shown should we repeat it
+	repeatAfter time.Duration
+}
+
+func (w *Warning) String() string {
+	return w.message
+}
+
+func (w *Warning) MarshalJSON() ([]byte, error) {
+	jw := jsonWarning{
+		Message:     w.message,
+		FirstAdded:  w.firstAdded,
+		LastAdded:   w.lastAdded,
+		ExpireAfter: w.expireAfter.String(),
+		RepeatAfter: w.repeatAfter.String(),
+	}
+	if !w.lastShown.IsZero() {
+		jw.LastShown = &w.lastShown
+	}
+
+	return json.Marshal(jw)
+}
+
+func (w *Warning) UnmarshalJSON(data []byte) error {
+	var jw jsonWarning
+	err := json.Unmarshal(data, &jw)
+	if err != nil {
+		return err
+	}
+	w.message = jw.Message
+	w.firstAdded = jw.FirstAdded
+	w.lastAdded = jw.LastAdded
+	if jw.LastShown != nil {
+		w.lastShown = *jw.LastShown
+	}
+	if jw.ExpireAfter != "" {
+		w.expireAfter, err = time.ParseDuration(jw.ExpireAfter)
+		if err != nil {
+			return err
+		}
+	}
+	if jw.RepeatAfter != "" {
+		w.repeatAfter, err = time.ParseDuration(jw.RepeatAfter)
+		if err != nil {
+			return err
+		}
+	}
+
+	return w.validate()
+}
+
+func (w *Warning) validate() (e error) {
+	if w.message == "" {
+		return errNoMessage
+	}
+	if strings.TrimSpace(w.message) != w.message {
+		return errBadMessage
+	}
+	if w.firstAdded.IsZero() {
+		return errNoFirstAdded
+	}
+	if w.expireAfter == 0 {
+		return errNoExpireAfter
+	}
+	if w.repeatAfter == 0 {
+		return errNoRepeatAfter
+	}
+	return nil
+}
+
+func (w *Warning) Expired(now time.Time) bool {
+	return w.lastAdded.Add(w.expireAfter).Before(now)
+}
+
+func (w *Warning) IsShowable(t time.Time) bool {
+	return (w.lastShown.IsZero() || w.lastShown.Add(w.repeatAfter).Before(t)) && !w.firstAdded.After(t)
+}
+
+// flattenWarning loops over the warnings map, and returns all
+// warnings therein as a flat list, for serialising.
+// Call with the lock held.
+func (s *State) flattenWarnings() []*Warning {
+	flat := make([]*Warning, 0, len(s.warnings))
+	for _, w := range s.warnings {
+		flat = append(flat, w)
+	}
+	return flat
+}
+
+// unflattenWarnings takes a flat list of warnings and replaces the
+// warning map with them.
+// Call with the lock held.
+func (s *State) unflattenWarnings(flat []*Warning) {
+	s.warnings = make(map[string]*Warning, len(flat))
+	for _, w := range flat {
+		s.warnings[w.message] = w
+	}
+}
+
+// AddWarning records a warning: if it's the first Warning with this
+// message it'll be added (with its firstAdded and lastAdded set to the
+// current time), otherwise the existing one will have its lastAdded
+// updated.
+func (s *State) AddWarning(message string) {
```

Comment:
I suggest "Warning" as the function name, and something a bit more comfortable:
```
func (s *State) Warning(format string, args ...interface{}) {
```

Reply:
I don't think we want warnings to have arguments like this. That is, we don't want "You only have 200MB of disk space left"; we want "Your disk space is running low". Otherwise the "only hold one warning of a given type" idea goes down the drain.

*If*, as an iteration, we move to use the `format` as the key, and have extra data in warnings where we store the args, then this would work. But I'd like to leave it as an iteration of the api (and call it `Warningf`? I'm bad with names)

`Warning` on its own seems strange to me; as it's a noun and not a verb I'd expect it to return a warning, not add it (like `state.Task(id)` returns a task). But we've established that I'm bad with names :man_shrugging:

Reply:
You are right, this should really be _Warn_ instead of _Warning_ (or Warnf if you want). Can we go with that? There's not much reason to wait.. it's just a Sprintf inside the function.

I agree with your point about being careful with arguments so we're not bothering people and losing the benefits of the system, but we do want arguments anyway so we can talk to people about specific things instead of just general messages. For example, colloquially:

 * Your snap "foo" is broken
 * Partition /foo is running low on space
 * Snapshot file at /.../foo is corrupted

etc.